File: src/extras-kio/libudsentry.cpp
#define WORKAROUND_INNER_CLASS_DEFINITION_KIO__UDSEntry
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <udsentry.h>
#include "libudsentry.h"
#include "libudsentry.hxx"

KIO__UDSEntry* KIO__UDSEntry_new() {
    return new KIO::UDSEntry();
}

KIO__UDSEntry* KIO__UDSEntry_new2(const KIO__UDSEntry* param1) {
    return new KIO::UDSEntry(*param1);
}

void KIO__UDSEntry_OperatorAssign(KIO__UDSEntry* self, const KIO__UDSEntry* param1) {
    self->operator=(*param1);
}

libqt_string KIO__UDSEntry_StringValue(const KIO__UDSEntry* self, unsigned int field) {
    QString _ret = self->stringValue(static_cast<uint>(field));
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

long long KIO__UDSEntry_NumberValue(const KIO__UDSEntry* self, unsigned int field) {
    return self->numberValue(static_cast<uint>(field));
}

bool KIO__UDSEntry_IsDir(const KIO__UDSEntry* self) {
    return self->isDir();
}

bool KIO__UDSEntry_IsLink(const KIO__UDSEntry* self) {
    return self->isLink();
}

void KIO__UDSEntry_Reserve(KIO__UDSEntry* self, int size) {
    self->reserve(static_cast<int>(size));
}

void KIO__UDSEntry_FastInsert(KIO__UDSEntry* self, unsigned int field, const libqt_string value) {
    QString value_QString = QString::fromUtf8(value.data, value.len);
    self->fastInsert(static_cast<uint>(field), value_QString);
}

void KIO__UDSEntry_FastInsert2(KIO__UDSEntry* self, unsigned int field, long long l) {
    self->fastInsert(static_cast<uint>(field), static_cast<long long>(l));
}

int KIO__UDSEntry_Count(const KIO__UDSEntry* self) {
    return self->count();
}

bool KIO__UDSEntry_Contains(const KIO__UDSEntry* self, unsigned int field) {
    return self->contains(static_cast<uint>(field));
}

libqt_list /* of unsigned int */ KIO__UDSEntry_Fields(const KIO__UDSEntry* self) {
    QList<unsigned int> _ret = self->fields();
    // Convert QList<> from C++ memory to manually-managed C memory
    unsigned int* _arr = static_cast<unsigned int*>(malloc(sizeof(unsigned int) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = _ret[i];
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void KIO__UDSEntry_Clear(KIO__UDSEntry* self) {
    self->clear();
}

void KIO__UDSEntry_Replace(KIO__UDSEntry* self, unsigned int field, const libqt_string value) {
    QString value_QString = QString::fromUtf8(value.data, value.len);
    self->replace(static_cast<uint>(field), value_QString);
}

void KIO__UDSEntry_Replace2(KIO__UDSEntry* self, unsigned int field, long long l) {
    self->replace(static_cast<uint>(field), static_cast<long long>(l));
}

long long KIO__UDSEntry_NumberValue2(const KIO__UDSEntry* self, unsigned int field, long long defaultValue) {
    return self->numberValue(static_cast<uint>(field), static_cast<long long>(defaultValue));
}

void KIO__UDSEntry_Delete(KIO__UDSEntry* self) {
    delete self;
}

File: include/extras-kio/libudsentry.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/kio-udsentry.html
pub const kio__udsentry = struct {
    /// New constructs a new KIO::UDSEntry object.
    ///
    ///
    pub fn New() QtC.KIO__UDSEntry {
        return qtc.KIO__UDSEntry_new();
    }

    /// New2 constructs a new KIO::UDSEntry object.
    ///
    /// ``` param1: QtC.KIO__UDSEntry ```
    pub fn New2(param1: ?*anyopaque) QtC.KIO__UDSEntry {
        return qtc.KIO__UDSEntry_new2(@ptrCast(param1));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#operator-eq)
    ///
    /// ``` self: QtC.KIO__UDSEntry, param1: QtC.KIO__UDSEntry ```
    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.KIO__UDSEntry_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#stringValue)
    ///
    /// ``` self: QtC.KIO__UDSEntry, field: u32, allocator: std.mem.Allocator ```
    pub fn StringValue(self: ?*anyopaque, field: u32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KIO__UDSEntry_StringValue(@ptrCast(self), @intCast(field));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kio::udsentry.StringValue: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#numberValue)
    ///
    /// ``` self: QtC.KIO__UDSEntry, field: u32 ```
    pub fn NumberValue(self: ?*anyopaque, field: u32) i64 {
        return qtc.KIO__UDSEntry_NumberValue(@ptrCast(self), @intCast(field));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#isDir)
    ///
    /// ``` self: QtC.KIO__UDSEntry ```
    pub fn IsDir(self: ?*anyopaque) bool {
        return qtc.KIO__UDSEntry_IsDir(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#isLink)
    ///
    /// ``` self: QtC.KIO__UDSEntry ```
    pub fn IsLink(self: ?*anyopaque) bool {
        return qtc.KIO__UDSEntry_IsLink(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#reserve)
    ///
    /// ``` self: QtC.KIO__UDSEntry, size: i32 ```
    pub fn Reserve(self: ?*anyopaque, size: i32) void {
        qtc.KIO__UDSEntry_Reserve(@ptrCast(self), @intCast(size));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#fastInsert)
    ///
    /// ``` self: QtC.KIO__UDSEntry, field: u32, value: []const u8 ```
    pub fn FastInsert(self: ?*anyopaque, field: u32, value: []const u8) void {
        const value_str = qtc.libqt_string{
            .len = value.len,
            .data = value.ptr,
        };
        qtc.KIO__UDSEntry_FastInsert(@ptrCast(self), @intCast(field), value_str);
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#fastInsert)
    ///
    /// ``` self: QtC.KIO__UDSEntry, field: u32, l: i64 ```
    pub fn FastInsert2(self: ?*anyopaque, field: u32, l: i64) void {
        qtc.KIO__UDSEntry_FastInsert2(@ptrCast(self), @intCast(field), @intCast(l));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#count)
    ///
    /// ``` self: QtC.KIO__UDSEntry ```
    pub fn Count(self: ?*anyopaque) i32 {
        return qtc.KIO__UDSEntry_Count(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#contains)
    ///
    /// ``` self: QtC.KIO__UDSEntry, field: u32 ```
    pub fn Contains(self: ?*anyopaque, field: u32) bool {
        return qtc.KIO__UDSEntry_Contains(@ptrCast(self), @intCast(field));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#fields)
    ///
    /// ``` self: QtC.KIO__UDSEntry, allocator: std.mem.Allocator ```
    pub fn Fields(self: ?*anyopaque, allocator: std.mem.Allocator) []u32 {
        const _arr: qtc.libqt_list = qtc.KIO__UDSEntry_Fields(@ptrCast(self));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(u32, _arr.len) catch @panic("kio::udsentry.Fields: Memory allocation failed");
        const _data: [*]u32 = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#clear)
    ///
    /// ``` self: QtC.KIO__UDSEntry ```
    pub fn Clear(self: ?*anyopaque) void {
        qtc.KIO__UDSEntry_Clear(@ptrCast(self));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#replace)
    ///
    /// ``` self: QtC.KIO__UDSEntry, field: u32, value: []const u8 ```
    pub fn Replace(self: ?*anyopaque, field: u32, value: []const u8) void {
        const value_str = qtc.libqt_string{
            .len = value.len,
            .data = value.ptr,
        };
        qtc.KIO__UDSEntry_Replace(@ptrCast(self), @intCast(field), value_str);
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#replace)
    ///
    /// ``` self: QtC.KIO__UDSEntry, field: u32, l: i64 ```
    pub fn Replace2(self: ?*anyopaque, field: u32, l: i64) void {
        qtc.KIO__UDSEntry_Replace2(@ptrCast(self), @intCast(field), @intCast(l));
    }

    /// [Qt documentation](https://api.kde.org/kio-udsentry.html#numberValue)
    ///
    /// ``` self: QtC.KIO__UDSEntry, field: u32, defaultValue: i64 ```
    pub fn NumberValue2(self: ?*anyopaque, field: u32, defaultValue: i64) i64 {
        return qtc.KIO__UDSEntry_NumberValue2(@ptrCast(self), @intCast(field), @intCast(defaultValue));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KIO__UDSEntry ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KIO__UDSEntry_Delete(@ptrCast(self));
    }
};

/// https://api.kde.org/kio-udsentry.html#types
pub const enums = struct {
    pub const ItemTypes = enum {
        pub const UDS_STRING: i32 = 16777216;
        pub const UDS_NUMBER: i32 = 33554432;
        pub const UDS_TIME: i32 = 100663296;
    };

    pub const StandardFieldTypes = enum {
        pub const UDS_SIZE: i32 = 33554433;
        pub const UDS_SIZE_LARGE: i32 = 33554434;
        pub const UDS_USER: i32 = 16777219;
        pub const UDS_ICON_NAME: i32 = 16777220;
        pub const UDS_GROUP: i32 = 16777221;
        pub const UDS_NAME: i32 = 16777222;
        pub const UDS_LOCAL_PATH: i32 = 16777223;
        pub const UDS_HIDDEN: i32 = 33554440;
        pub const UDS_ACCESS: i32 = 33554441;
        pub const UDS_MODIFICATION_TIME: i32 = 100663306;
        pub const UDS_ACCESS_TIME: i32 = 100663307;
        pub const UDS_CREATION_TIME: i32 = 100663308;
        pub const UDS_FILE_TYPE: i32 = 33554445;
        pub const UDS_LINK_DEST: i32 = 16777230;
        pub const UDS_URL: i32 = 16777231;
        pub const UDS_MIME_TYPE: i32 = 16777232;
        pub const UDS_GUESSED_MIME_TYPE: i32 = 16777233;
        pub const UDS_XML_PROPERTIES: i32 = 16777234;
        pub const UDS_EXTENDED_ACL: i32 = 33554451;
        pub const UDS_ACL_STRING: i32 = 16777236;
        pub const UDS_DEFAULT_ACL_STRING: i32 = 16777237;
        pub const UDS_DISPLAY_NAME: i32 = 16777238;
        pub const UDS_TARGET_URL: i32 = 16777239;
        pub const UDS_DISPLAY_TYPE: i32 = 16777240;
        pub const UDS_ICON_OVERLAY_NAMES: i32 = 16777241;
        pub const UDS_COMMENT: i32 = 16777242;
        pub const UDS_DEVICE_ID: i32 = 33554459;
        pub const UDS_INODE: i32 = 33554460;
        pub const UDS_RECURSIVE_SIZE: i32 = 33554461;
        pub const UDS_LOCAL_USER_ID: i32 = 33554462;
        pub const UDS_LOCAL_GROUP_ID: i32 = 33554463;
        pub const UDS_EXTRA: i32 = 16777316;
        pub const UDS_EXTRA_END: i32 = 16777356;
    };
};
